src/nvim/eval.c

```diff
@@ -18143,27 +18172,42 @@ static void f_writefile(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   bool binary = false;
   bool append = false;
   if (argvars[2].v_type != VAR_UNKNOWN) {
-    if (vim_strchr(get_tv_string(&argvars[2]), 'b')) {
+    const char *const flags = (const char *)get_tv_string_chk(&argvars[2]);
+    if (flags == NULL) {
+      return;
+    }
+    if (strchr(flags, 'b')) {
       binary = true;
     }
-    if (vim_strchr(get_tv_string(&argvars[2]), 'a')) {
+    if (strchr(flags, 'a')) {
       append = true;
     }
   }
 
-  // Always open the file in binary mode, library functions have a mind of
-  // their own about CR-LF conversion.
-  char_u *fname = get_tv_string(&argvars[1]);
-  FILE *fd;
-  if (*fname == NUL || (fd = mch_fopen((char *)fname,
-       append ? APPENDBIN : WRITEBIN)) == NULL) {
-    EMSG2(_(e_notcreate), *fname == NUL ? (char_u *)_("<empty>") : fname);
-    rettv->vval.v_number = -1;
+  const char buf[NUMBUFLEN];
+  const char *const fname = (const char *)get_tv_string_buf_chk(&argvars[1],
+                                                                (char_u *)buf);
+  if (fname == NULL) {
+    return;
+  }
+  FileDescriptor *fp;
+  int error;
+  rettv->vval.v_number = -1;
+  if (*fname == NUL) {
+    EMSG(_("E482: Can't open file with an empty name"));
+  } else if ((fp = file_open_new(&error, fname,
+                                 ((append ? kFileAppend : kFileTruncate)
+                                  | kFileCreate), 0666)) == NULL) {
+    emsgf(_("E482: Can't open file %s for writing: %s"),
+          fname, os_strerror(error));
   } else {
-    if (write_list(fd, argvars[0].vval.v_list, binary) == false) {
-      rettv->vval.v_number = -1;
+    if (write_list(fp, argvars[0].vval.v_list, binary)) {
+      rettv->vval.v_number = 0;
```

Comment:
This could be too early to signal a successful write. Flushing of the remaining `RBuffer` could fail in `file_free()` and writing to the file could be incomplete.